File: camera.py
import numpy as np
import math
from pyglm import glm

from OpenGL.GL import *

class Camera:
    def __init__(self, data):
        yaw, pitch, zoom = data
        self.yaw = yaw
        self.pitch = pitch
        self.zoom = zoom

    def getFacing(self):
        y = math.sin(self.pitch)
        x = math.cos(self.yaw) * math.cos(self.pitch)
        z = math.sin(self.yaw) * math.cos(self.pitch)
        self.facing = (x, y, z)        

    def orbit(self, yaw, pitch):
        self.yaw += yaw
        self.pitch += pitch

        if self.pitch > math.radians(90):
            self.pitch = math.radians(90)
        if self.pitch < math.radians(-90):
            self.pitch = math.radians(-90)
    
    def _zoom(self, zoom):
        self.zoom += zoom
        if self.zoom <= 0:
            self.zoom = 0
    
    def calculateViewMatrix(self, modelViewLocation):
        # Facing unit vector
        self.getFacing()
        xc, yc, zc = self.facing

        # Scale facing vector
        x = -self.zoom * xc
        y = -self.zoom * yc
        z = -self.zoom * zc

        self.position = glm.vec3(x, y, z)
        eye = self.position
        target = glm.vec3(0.0, 0.0, 0.0)
        up = glm.vec3(0.0, 1.0, 0.0)
        self.view_transform = glm.lookAt(eye, target, up)
        glUniformMatrix4fv(modelViewLocation, 1, GL_TRUE, np.array(self.view_transform))

File: cpp_engine/src/camera.cpp
#include "camera.h"

Camera::Camera(float yaw, float pitch, float zoom) 
    : yaw(yaw), pitch(pitch), zoom(zoom) {}

glm::vec3 Camera::getFacing() {
    glm::vec3 facing;
    facing.x = cos(glm::radians(yaw)) * cos(glm::radians(pitch));
    facing.y = sin(glm::radians(pitch));
    facing.z = sin(glm::radians(yaw)) * cos(glm::radians(pitch));
    return glm::normalize(facing);
}

glm::mat4 Camera::getViewMatrix() {
    return viewMatrix;
}

void Camera::updateViewMatrix(Shader &shader) {
    position = -zoom * getFacing();
    viewMatrix = glm::lookAt(position, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f));
    shader.setMat4("view", viewMatrix);
}

void Camera::orbit(float xoffset, float yoffset) {
    yaw += xoffset;
    pitch += yoffset;

    if (pitch > 89.0f)  pitch = 89.0f;
    if (pitch < -89.0f) pitch = -89.0f;
}

void Camera::changeZoom(float yoffset) {
    zoom -= yoffset;
    if (zoom < 1.0f) zoom = 1.0f;
}